File: frontend/digital-library/src/types/library.ts
// frontend/digital-library/src/types/library.ts

export interface Faculty {
  id: number;
  code: string;
  name: string;
  document_count: number;
  subject_count: number;
}

export interface Subject {
  id: number;
  faculty_id: number;
  code: string;
  name: string;
  description?: string | null;
  document_count: number;
  avg_rating: number;
  created_at?: string | null;
}

export interface CommunityDocument {
  id: number;
  title: string;
  description?: string | null;
  file_type?: string | null;
  file_size?: number | null;
  subject_id?: number | null;
  subject_name?: string | null;
  faculty_id?: number | null;
  faculty_name?: string | null;
  doc_type?: string | null;
  academic_year?: string | null;
  owner_id: number;
  contributed_by: string;
  rating_avg: number;
  rating_count: number;
  download_count: number;
  created_at?: string | null;
}

export interface Rating {
  id: number;
  document_id: number;
  user_id: number;
  user_name: string;
  stars: number;
  comment?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface Submission {
  id: number;
  source_document_id: number;
  source_document_title: string;
  published_document_id?: number | null;
  submitter_id: number;
  submitter_name: string;
  subject_id: number;
  subject_name: string;
  faculty_name: string;
  academic_year?: string | null;
  doc_type?: string | null;
  note?: string | null;
  status: 'pending' | 'approved' | 'rejected';
  reject_reason?: string | null;
  reviewed_by?: number | null;
  reviewed_at?: string | null;
  created_at?: string | null;
}

export interface SubmitPayload {
  source_document_id: number;
  subject_id: number;
  academic_year?: string;
  doc_type?: string;
  note?: string;
}

export interface PaginatedCommunityDocuments {
  items: CommunityDocument[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface PaginatedSubmissions {
  items: Submission[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

File: backend/app/schemas/library.py
# backend/app/schemas/library.py
from datetime import datetime
from typing import Optional, List, Dict
from pydantic import BaseModel, ConfigDict, Field

class FacultyLibraryOut(BaseModel):
    id: int
    code: str
    name: str
    document_count: int = 0
    subject_count: int = 0
    model_config = ConfigDict(from_attributes=True)

class SubjectOut(BaseModel):
    id: int
    faculty_id: int
    code: str
    name: str
    description: Optional[str] = None
    document_count: int = 0
    avg_rating: float = 0.0
    created_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)

class RatingCreate(BaseModel):
    stars: int = Field(..., ge=1, le=5, description="Điểm đánh giá từ 1 đến 5 sao")
    comment: Optional[str] = None

class RatingUpdate(BaseModel):
    stars: Optional[int] = Field(None, ge=1, le=5)
    comment: Optional[str] = None

class RatingOut(BaseModel):
    id: int
    document_id: int
    user_id: int
    user_name: str
    stars: int
    comment: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)

class SubmissionCreate(BaseModel):
    source_document_id: int
    subject_id: int
    academic_year: Optional[str] = None
    doc_type: Optional[str] = None
    note: Optional[str] = None

class SubmissionRejectPayload(BaseModel):
    reason: str = Field(..., min_length=5, description="Lý do từ chối tài liệu")

class SubmissionOut(BaseModel):
    id: int
    source_document_id: int
    source_document_title: str
    published_document_id: Optional[int] = None
    submitter_id: int
    submitter_name: str
    subject_id: int
    subject_name: str
    faculty_name: str
    academic_year: Optional[str] = None
    doc_type: Optional[str] = None
    note: Optional[str] = None
    status: str  # pending | approved | rejected
    reject_reason: Optional[str] = None
    reviewed_by: Optional[int] = None
    reviewed_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)

class CommunityDocumentOut(BaseModel):
    id: int
    title: str
    description: Optional[str] = None
    file_type: Optional[str] = None
    file_size: Optional[int] = None
    subject_id: Optional[int] = None
    subject_name: Optional[str] = None
    faculty_id: Optional[int] = None
    faculty_name: Optional[str] = None
    doc_type: Optional[str] = None
    academic_year: Optional[str] = None
    owner_id: int
    contributed_by: str
    rating_avg: float = 0.0
    rating_count: int = 0
    download_count: int = 0
    created_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)

class PaginatedCommunityDocuments(BaseModel):
    items: List[CommunityDocumentOut]
    total: int
    page: int
    page_size: int
    total_pages: int

class PaginatedSubmissions(BaseModel):
    items: List[SubmissionOut]
    total: int
    page: int
    page_size: int
    total_pages: int
